### PR TITLE
Band Protocol Grant Application

### PR DESCRIPTION
**Pull request by buranapongb**

### CODE REVIEW

**Comment by TtomaS7** [comment]
Thank you for submitting your application to ink!ubator. We appreciate your interest and the effort you've put into your proposal.
As I review your application, we have a couple of queries that we'd like to clarify in order to better understand your project:

You've provided a project timeline estimation ranging from 2-4 months. The span of this timeframe is quite broad. Could you please specify more clearly, or perhaps provide a more narrowed down estimate?

The presented cost - does it account for the monitoring costs for any of the chains? If it does, could you please indicate for which chains and the duration this cost covers?

We're eager to learn more about your proposal, and your responses to these questions will greatly assist us in our assessment.
Thank you once again for your interest in ink!ubator!!

**Comment by buranapongb** [comment]
> Thank you for submitting your application to ink!ubator. We appreciate your interest and the effort you've put into your proposal. As I review your application, we have a couple of queries that we'd like to clarify in order to better understand your project:
> 
> You've provided a project timeline estimation ranging from 2-4 months. The span of this timeframe is quite broad. Could you please specify more clearly, or perhaps provide a more narrowed down estimate?
> 
> The presented cost - does it account for the monitoring costs for any of the chains? If it does, could you please indicate for which chains and the duration this cost covers?
> 
> We're eager to learn more about your proposal, and your responses to these questions will greatly assist us in our assessment. Thank you once again for your interest in ink!ubator!!

**Comment by buranapongb** [comment]
@TtomaS7

**Comment by TtomaS7** [comment]
Thanks for clarifications! 
I will assign curators for the review.

File: applications/band-grants.md [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+- Data models / API specifications of the core functionality
+- An overview of the technology stack to be used
+- Documentation of core components, protocols, architecture, etc. to be deployed
+- PoC/MVP or other relevant prior work or research on the topic
+- What your project is _not_ or will _not_ provide or implement
+  - This is a place for you to manage expectations and to clarify any limitations that might not be obvious
+- Do you need an audit for the contacts? **YES/NO**
+(don’t add it as part of Milestones! The auditor will be chosen among Auditor track participants)
+- CATEGORY: Infrastructure OR Canary Dapp OR Technical Showcase
```
Comment (h4x3rotab) [inline applications/band-grants.md:48]: Could you elaborate on the questions above?

File: applications/band-grants.md [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+| 2. | Oracle Script | An oracle script on BandChain which will verify and aggregate data from the data sources on BandChain|
+| 3. | BandChain Requester | A service that will prepare and request data from BandChain to be sent to Astar | 
+| 4. | Astar ink! Relayer | A service that will prepare and execute data to be bridged over to Astar|
+| 5. | ink! Signer | A seperate but crucial service that will used concurrently to Astar to send over data |
+| 6. | PDS | A service on BandChain that will allow paid services to deliver data to BandChain safely and securely |
+| 7. | Smart contract | An ink! smart contract that will allow consumers to query various price feed from BandChain|
+| 8. | Testing and Audit | Testing|
+| 9. | Maintenance | Reviewer can continuously read the state of the contract to get an asset's price|
+| 10. | Infrastructure | Hosting cost - Storage, GCP|
+| 11. | Monitoring | 24/7 monitoring service ensures the continuous operation of the relayer. This includes the comprehensive monitoring of both the Astar EVM and Astar WASM chains for a duration of one year|
```
Comment (h4x3rotab) [inline applications/band-grants.md:146]: Do you expect to expand the oracle service to other ink! based blockchain like AlephZero?

File: applications/band-grants.md [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+- An indication of how your project relates to ink! ecosystem.
+    - Band Protocol has the potential to become a crucial infrastructure provider for the ink! Ecosystem. As projects develop on Astar, they will require reliable oracle services to deliver precise real-world data, and VRF to produce randomized results in their gaming and lottery mechanisms. By integrating with Astar, we can prepare and offer these essential services to Dapps building on the platform. Additionally, developers can be assured that they will receive high-quality data from a team of experienced professionals managing the price feed. By leveraging Band Protocol's expertise, Astar can enhance its offerings and provide a more comprehensive, reliable ecosystem for its users. Furthermore, we have a strong track record of collaborating with and offering oracle solutions for astar for more than a year now. This extensive experience positions us as a knowledgeable and skilled team in supporting the polkadot ecosystem and we are truly enthusiastic about continuing our involvement to its growth.
```
Comment (h4x3rotab) [inline applications/band-grants.md:22]: Does Band plan to provide the full access to the standard dataset as well as the VRF features to ink contracts this time?

File: applications/band-grants.md [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+| **0b.** | Documentation | We will provide inline documentation of the code, a basic tutorial on how to use the smart contract and an example consumer contract that shows a practical example on how to use the smart contract|
+| **0c.** | Testing and Testing Guide | Core functions will be fully covered by comprehensive unit tests.|
+| **0d.** | Docker | N/A for smart contracts|
+| 0e. | Article | We will publish an **article**/workshop that explains [...] (what was done/achieved as part of the grant). (Content, language and medium should reflect your target audience described above.) |
+| 1. | Data Source | Data sources on BandChain which will be the core of getting data that will be delivered to Astar |
+| 2. | Oracle Script | An oracle script on BandChain which will verify and aggregate data from the data sources on BandChain|
+| 3. | BandChain Requester | A service that will prepare and request data from BandChain to be sent to Astar | 
+| 4. | Astar ink! Relayer | A service that will prepare and execute data to be bridged over to Astar|
+| 5. | ink! Signer | A seperate but crucial service that will used concurrently to Astar to send over data |
```
Comment (h4x3rotab) [inline applications/band-grants.md:140]: What's the relationship between the relayer and the signer? Could you elaborate on the signing process, i.e. how is the security model? Does it use threshold signature to create efficient transactions?

**Review by h4x3rotab — COMMENTED** [review]
Excited to see infrastructure project like oracle in the inkubator program application. I've left a few questions inline. Please check. :D

**Review by h4x3rotab — COMMENTED** [review]
Excited to see infrastructure project like oracle in the inkubator program application. I've left a few questions inline. Please check. :D

File: applications/band-grants.md (hunk 1 above)
Comment (buranapongb) [reply applications/band-grants.md]: Provided the missing info on the updated doc.

File: applications/band-grants.md (hunk 2 above)
Comment (buranapongb) [reply applications/band-grants.md:151]: We are excited about the opportunity to expand our oracle service to additional ink!-based blockchains. However, it is important to clarify that the grant we are seeking will be exclusively allocated to support the development and maintenance of Astar WASM.

File: applications/band-grants.md (hunk 3 above)
Comment (buranapongb) [reply applications/band-grants.md:22]: We have the capability to offer up to 15 symbols from our standard dataset (https://data.bandprotocol.com/). The VRF will not be included in ink contracts at this time. Nevertheless, we are eager and willing to integrate our VRF into ink contracts in the near future.

**Comment by h4x3rotab** [comment]
Sorry for the late reply. I believe this is a strong application that benefits the ink ecosystem a lot. Just the last question: I see there's only one milestone defined in this proposal. So you will need to finish the full project and get the bounty for once. Just want to confirm if it's expected.

File: applications/band-grants.md (hunk 2 above)
Comment (buranapongb) [reply applications/band-grants.md:159]: Updated, see changes in line 56

**Comment by buranapongb** [comment]
> Sorry for the late reply. I believe this is a strong application that benefits the ink ecosystem a lot. Just the last question: I see there's only one milestone defined in this proposal. So you will need to finish the full project and get the bounty for once. Just want to confirm if it's expected.

Is it feasible to divide the grant into multiple phases? For instance, we could consider a two-phase approach: 20% upfront and the remaining amount upon task completion. Alternatively, we could establish smaller milestones, although we currently find that approach inconvenient.

**Comment by h4x3rotab** [comment]
> > Sorry for the late reply. I believe this is a strong application that benefits the ink ecosystem a lot. Just the last question: I see there's only one milestone defined in this proposal. So you will need to finish the full project and get the bounty for once. Just want to confirm if it's expected.
> 
> Is it feasible to divide the grant into multiple phases? For instance, we could consider a two-phase approach: 20% upfront and the remaining amount upon task completion. Alternatively, we could establish smaller milestones, although we currently find that approach inconvenient.

I don't think upfront payment is possible, but it's good to split the big milestone to multiple smaller milestones. Just note that each milestone should have clear deliverables.

**Comment by buranapongb** [comment]
> > > Sorry for the late reply. I believe this is a strong application that benefits the ink ecosystem a lot. Just the last question: I see there's only one milestone defined in this proposal. So you will need to finish the full project and get the bounty for once. Just want to confirm if it's expected.
> > 
> > 
> > Is it feasible to divide the grant into multiple phases? For instance, we could consider a two-phase approach: 20% upfront and the remaining amount upon task completion. Alternatively, we could establish smaller milestones, although we currently find that approach inconvenient.
> 
> I don't think upfront payment is possible, but it's good to split the big milestone to multiple smaller milestones. Just note that each milestone should have clear deliverables.

If we split the big milestone into smaller milestones, will the grant award be distributed evenly? Say we split into two tasks, will we receive 50% of the grant award we asked for right after we've completed each task?

**Comment by h4x3rotab** [comment]
The amount of the number should be justified by the workload of each milestone. So you can split the big one to smaller ones, and propose the amount according to the workload.

**Comment by buranapongb** [comment]
> The amount of the number should be justified by the workload of each milestone. So you can split the big one to smaller ones, and propose the amount according to the workload.

We've just split our task into two, please have a look and let us know if you have any comments!

**Comment by h4x3rotab** [comment]
> > The amount of the number should be justified by the workload of each milestone. So you can split the big one to smaller ones, and propose the amount according to the workload.
> 
> We've just split our task into two, please have a look and let us know if you have any comments!

You might have forgot to push the latest update. Could you take a look? So far it only has one milestone.

**Comment by buranapongb** [comment]
> > > The amount of the number should be justified by the workload of each milestone. So you can split the big one to smaller ones, and propose the amount according to the workload.
> > 
> > 
> > We've just split our task into two, please have a look and let us know if you have any comments!
> 
> You might have forgot to push the latest update. Could you take a look? So far it only has one milestone.

My apologies, it should be updated now.

**Review by DoubleOTheven — APPROVED** [review]

File: applications/band-grants.md [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+| 2. | Oracle Script | An oracle script on BandChain which will verify and aggregate data from the data sources on BandChain|
+| 3. | PDS | A service on BandChain that will allow paid services to deliver data to BandChain safely and securely |
+| 4. | Smart contract | An ink! smart contract that will allow consumers to query various price feed from BandChain|
+| 5. | Testing and Audit | Testing|
+### Milestone 2
+- **Estimated duration:** 1-2 months
+- **FTE:**  2
+- **Costs:** 35,000 USD
+
+> :exclamation: **The default deliverables 0a-0d below are mandatory for all milestones**, and deliverable 0e at least for the last one.
```
Comment (HCastano) [inline applications/band-grants.md:156]:
```suggestion
```

File: applications/band-grants.md [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+
+- Band Protocol has been a trusted player in the blockchain space since 2017. We have experienced the ups and downs of the industry and have consistently performed well. On the technical side, our engineers have extensive experience integrating with EVM, WASM, and Rust-based projects, making us well-equipped to deliver a high-quality final product. Notably, we have successfully integrated with several prominent projects in the Cosmos ecosystem, including Injective, Cronos, OKC, and Secret Network. Our expertise and track record of successful integrations make us a reliable partner for projects looking to incorporate decentralized data oracles into their architecture.
+
+### Team Code Repos
+
+- https://github.com/bandprotocol/
+
+Please also provide the GitHub accounts of all team members. If they contain no activity, references to projects hosted elsewhere or live are also fine.
+
+- https://github.com/warittornc
```
Comment (HCastano) [inline applications/band-grants.md:100]: Is there a GitHub profile for the other FTE you are asking for funding for?

File: applications/band-grants.md [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+- CATEGORY: Infrastructure OR Canary Dapp OR Technical Showcase
+    - Infrastructure
+
+
+Things that shouldn’t be part of the application:
+- The (future) tokenomics of your project 
+- For non-infrastructure projects—deployment and hosting costs, maintenance or audits
+- Business-oriented activities (marketing, business planning), events or outreach
+
+
```
Comment (HCastano) [inline applications/band-grants.md:69]:
```suggestion
```

File: applications/band-grants.md [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+### Milestone 2
+- **Estimated duration:** 1-2 months
+- **FTE:**  2
+- **Costs:** 35,000 USD
+
+> :exclamation: **The default deliverables 0a-0d below are mandatory for all milestones**, and deliverable 0e at least for the last one. 
+
+|  Number | Deliverable | Specification |
+| -----: | ----------- | ------------- |
+| **0a.** | License | Apache 2.0 / GPLv3 / MIT / Unlicense |
```
Comment (HCastano) [inline applications/band-grants.md:160]: Same thing here

**Review by HCastano — COMMENTED** [review]
There's still a few things to clean up, but it's getting there

**Comment by HCastano** [comment]
@buranapongb any updates to my questions 😄 ?

**Review by 0xMarkian — APPROVED** [review]

**Comment by HCastano** [comment]
@buranapongb pinging again

**Comment by buranapongb** [comment]
> @buranapongb pinging again

@HCastano I've updated the file, please let me know if there are any other comments.

File: applications/band-grants.md [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+- **Costs:** 45,000 USD
+
+> :exclamation: **The default deliverables 0a-0d below are mandatory for all milestones**, and deliverable 0e at least for the last one. 
+
+|  Number | Deliverable | Specification |
+| -----: | ----------- | ------------- |
+| **0a.** | License | Apache 2.0 / GPLv3 / MIT / Unlicense |
+| **0b.** | Documentation | We will provide inline documentation of the code, a basic tutorial on how to use the smart contract and an example consumer contract that shows a practical example on how to use the smart contract|
+| **0c.** | Testing and Testing Guide | Core functions will be fully covered by comprehensive unit tests.|
+| **0d.** | Docker | N/A for smart contracts|
```
[the remark this answers is not in the document]
Comment (HCastano) [reply applications/band-grants.md:126]: I would like to see a Docker image provided here

File: applications/band-grants.md [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+| 0e. | Article | We will publish an **article**/workshop that explains [...] (what was done/achieved as part of the grant). (Content, language and medium should reflect your target audience described above.) |
+| 1. | Data Source | Data sources on BandChain which will be the core of getting data that will be delivered to Astar |
+| 2. | Oracle Script | An oracle script on BandChain which will verify and aggregate data from the data sources on BandChain|
+| 3. | PDS | A service on BandChain that will allow paid services to deliver data to BandChain safely and securely |
+| 4. | Smart contract | An ink! smart contract that will allow consumers to query various price feed from BandChain |
+| 5. | Testing and Audit | Unit tests for each component and live testing and audits for the ink! smart contract on testnet |
```
Comment (HCastano) [inline applications/band-grants.md:132]: The audit wouldn't be part of this milestone.

You should have unit tests and some sort of automated integration test suite

File: applications/band-grants.md [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+| **0d.** | Docker | N/A for smart contracts|
+| 0e. | Article | We will publish an **article**/workshop that explains [...] (what was done/achieved as part of the grant). (Content, language and medium should reflect your target audience described above.) |
+| 1. | BandChain Requester | A service that will prepare and request data from BandChain to be sent to Astar | 
+| 2. | Astar ink! Relayer | A service that will prepare and execute data to be bridged over to Astar|
+| 3. | ink! Signer | A seperate service which upon request, creates and signs a transction which is sent to the relayer service. The signer service is run on an isolated VPC to prevent unauthorized requests and uses a key management service to keep the signing key secure |
+| 4. | Maintenance | Reviewer can continuously read the state of the contract to get an asset's price|
+| 5. | Infrastructure | Hosting cost - Storage, GCP|
+| 6. | Monitoring | 24/7 monitoring service ensures the continuous operation of the relayer. This includes the comprehensive monitoring of both the Astar EVM and Astar WASM chains for a duration of one year|
```
Comment (HCastano) [inline applications/band-grants.md:150]: Why does the Astar EVM chain need to be monitored?

What tools are you planning to use here?

**Comment by TtomaS7** [comment]
Congratulations! We are thrilled to officially welcome you to the Ink!ubator program!! We can't wait to see the remarkable contributions you'll bring to the ecosystem.
In terms of the next steps, we are eagerly anticipating your first delivery, which is due in 2-3 months. When it’s ready, please create a PR here - https://github.com/use-inkubator/Grant-Milestone-Delivery

If you have any questions, concerns, or if you just want to brainstorm ideas, don't hesitate to reach out. We encourage you to utilize our Element channel for any inquiries - https://matrix.to/#/#sm-bounty-support:matrix.org

As you join the Ink!ubator family, please remember that you are not alone on this journey. We are here to support you every step of the way. We believe in your potential and we can't wait to see where this journey takes us all.

Congratulations once again and welcome aboard!

**Pull request merged**